--- a/libipt_FULLCONENAT.c
+++ b/libipt_FULLCONENAT.c
@@ -1,3 +1,7 @@
+#if !defined(__GLIBC__)
+/* avoid IPPROTO_IP,in_addr,in6_addr,sockaddr_in6,etc. redefine */
+#define _LIBC_COMPAT_H
+#endif
 #include <stdio.h>
 #include <netdb.h>
 #include <string.h>
--- a/libip6t_FULLCONENAT.c
+++ b/libip6t_FULLCONENAT.c
@@ -1,3 +1,7 @@
+#if !defined(__GLIBC__)
+/* avoid IPPROTO_IP,in_addr,in6_addr,sockaddr_in6,etc. redefine */
+#define _LIBC_COMPAT_H
+#endif
 #include <stdio.h>
 #include <netdb.h>
 #include <string.h>
