--- a/system-linux.c
+++ b/system-linux.c
@@ -1720,7 +1720,7 @@
 {
 	return system_link_del(vlandev->ifname);
 }
-
+#if 0
 static void
 system_set_ethtool_settings(struct device *dev, struct device_settings *s)
 {
@@ -1771,7 +1771,7 @@
 	ecmd.cmd = ETHTOOL_SSET;
 	ioctl(sock_ioctl, SIOCETHTOOL, &ifr);
 }
-
+#endif
 void
 system_if_get_settings(struct device *dev, struct device_settings *s)
 {
@@ -1995,7 +1995,7 @@
 		system_set_drop_unsolicited_na(dev, s->drop_unsolicited_na ? "1" : "0");
 	if (apply_mask & DEV_OPT_ARP_ACCEPT)
 		system_set_arp_accept(dev, s->arp_accept ? "1" : "0");
-	system_set_ethtool_settings(dev, s);
+	//system_set_ethtool_settings(dev, s);
 }
 
 int system_if_up(struct device *dev)
