From a84636fa28275f159910a7b8d2e54cbd5358edf8 Mon Sep 17 00:00:00 2001
From: Liangbin Lian <jjm2473@gmail.com>
Date: Tue, 19 Dec 2023 14:14:57 +0800
Subject: [PATCH] rtl8366_smi: fix CE on debugfs enabled

---
 drivers/net/phy/rtl8366_smi.c | 14 ++------------
 1 file changed, 2 insertions(+), 12 deletions(-)

diff --git a/drivers/net/phy/rtl8366_smi.c b/drivers/net/phy/rtl8366_smi.c
index e115b28c3..13f97f77e 100644
--- a/drivers/net/phy/rtl8366_smi.c
+++ b/drivers/net/phy/rtl8366_smi.c
@@ -944,13 +944,8 @@ static void rtl8366_debugfs_init(struct rtl8366_smi *smi)
 	}
 	root = smi->debugfs_root;
 
-	node = debugfs_create_x16("reg", S_IRUGO | S_IWUSR, root,
+	debugfs_create_x16("reg", S_IRUGO | S_IWUSR, root,
 				  &smi->dbg_reg);
-	if (!node) {
-		dev_err(smi->parent, "Creating debugfs file '%s' failed\n",
-			"reg");
-		return;
-	}
 
 	node = debugfs_create_file("val", S_IRUGO | S_IWUSR, root, smi,
 				   &fops_rtl8366_regs);
@@ -968,13 +963,8 @@ static void rtl8366_debugfs_init(struct rtl8366_smi *smi)
 		return;
 	}
 
-	node = debugfs_create_u8("vlan_4k_page", S_IRUGO | S_IWUSR, root,
+	debugfs_create_u8("vlan_4k_page", S_IRUGO | S_IWUSR, root,
 				  &smi->dbg_vlan_4k_page);
-	if (!node) {
-		dev_err(smi->parent, "Creating debugfs file '%s' failed\n",
-			"vlan_4k_page");
-		return;
-	}
 
 	node = debugfs_create_file("vlan_4k", S_IRUSR, root, smi,
 				   &fops_rtl8366_vlan_4k);
-- 
2.31.0

