From ca4343eb308b019c6bc6930361b56306cd3ba55d Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Mon, 5 Jun 2023 18:48:45 +0800
Subject: [PATCH] fix boorargs not in effect

---
 arch/arm/mach-rockchip/board.c | 2 ++
 1 file changed, 2 insertions(+)

--- a/arch/arm/mach-rockchip/board.c
+++ b/arch/arm/mach-rockchip/board.c
@@ -1159,6 +1159,7 @@ int board_rng_seed(struct abuf *buf)
 	return 0;
 }
 
+#ifdef CONFIG_CMD_BOOT_ANDROID
 /*
  * Pass fwver when any available.
  */
@@ -1384,6 +1385,7 @@ char *board_fdt_chosen_bootargs(void *fd
 
 	return (char *)bootargs;
 }
+#endif
 
 int ft_verify_fdt(void *fdt)
 {
