From 426af793bd894b619db9a1d529d8441cb1a01c26 Mon Sep 17 00:00:00 2001
From: Liangbin Lian <jjm2473@gmail.com>
Date: Mon, 7 Mar 2022 16:25:57 +0800
Subject: [PATCH] enable sata

---
 include/configs/rtd1295_qa_emmc_64.h | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/include/configs/rtd1295_qa_emmc_64.h b/include/configs/rtd1295_qa_emmc_64.h
index e7ead65a5f..dc1340330f 100644
--- a/include/configs/rtd1295_qa_emmc_64.h
+++ b/include/configs/rtd1295_qa_emmc_64.h
@@ -90,6 +90,20 @@
 	#define CONFIG_LZMA
 	#define CONFIG_CMD_LZMADEC
 
+	/* SATA */
+	#define CONFIG_RTK_AHSATA
+	#define CONFIG_CMD_SATA
+
+	#ifdef CONFIG_CMD_SATA
+		#define CONFIG_RTK_AHSATA
+		#define CONFIG_SYS_SATA_MAX_DEVICE      1
+		#define CONFIG_DWC_AHSATA_BASE_ADDR     0x9803F000
+		#define CONFIG_LBA48
+		#define CONFIG_SYS_64BIT_LBA
+		#define CONFIG_LIBATA
+		#define CONFIG_PORT0_POWER_PIN			56
+	#endif
+
 #elif defined(CONFIG_SYS_RTK_NAND_FLASH)
 	/* Flash writer setting:
 	*   The corresponding setting will be located at
-- 
2.31.0

