From b781c53acc25fc6108ddd4577b61b281f71ea7d2 Mon Sep 17 00:00:00 2001
From: jjm2473 <jjm2473@gmail.com>
Date: Tue, 25 Feb 2025 13:49:38 +0800
Subject: [PATCH] rknpu: fix config_clks_t in include/linux/pm_opp.h changed by
 openwrt

---
 drivers/rknpu/rknpu_devfreq.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/rknpu/rknpu_devfreq.c b/drivers/rknpu/rknpu_devfreq.c
index d8db3a3..e67cf4e 100644
--- a/drivers/rknpu/rknpu_devfreq.c
+++ b/drivers/rknpu/rknpu_devfreq.c
@@ -266,9 +266,16 @@ static int npu_opp_config_regulators(struct device *dev,
 					      count, &rknpu_dev->opp_info);
 }
 
+#ifndef BUILD_OPENWRT
 static int npu_opp_config_clks(struct device *dev, struct opp_table *opp_table,
 			       struct dev_pm_opp *opp, void *data,
 			       bool scaling_down)
+#else
+static int npu_opp_config_clks(struct device *dev, struct opp_table *opp_table,
+			       struct dev_pm_opp *old_opp,
+			       struct dev_pm_opp *opp, void *data,
+			       bool scaling_down)
+#endif
 {
 	struct rknpu_device *rknpu_dev = dev_get_drvdata(dev);
 
-- 
2.51.0

