From 09acb07f274a29341c32c09b932926aece67ebf4 Mon Sep 17 00:00:00 2001
From: Liangbin Lian <jjm2473@gmail.com>
Date: Fri, 13 Jun 2025 10:34:51 +0800
Subject: [PATCH] disable efi loader

---
 lib/efi_loader/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/efi_loader/Kconfig b/lib/efi_loader/Kconfig
index 6ffefa9103f..81f4711b5cf 100644
--- a/lib/efi_loader/Kconfig
+++ b/lib/efi_loader/Kconfig
@@ -11,7 +11,7 @@ config EFI_LOADER
 	# We need EFI_STUB_32BIT to be set on x86_32 with EFI_STUB
 	depends on !EFI_STUB || !X86 || X86_64 || EFI_STUB_32BIT
 	depends on !EFI_APP
-	default y if !ARM || SYS_CPU = armv7 || SYS_CPU = armv8
+	# default y if !ARM || SYS_CPU = armv7 || SYS_CPU = armv8
 	select BLK
 	select CHARSET
 	# We need to send DM events, dynamically, in the EFI block driver
-- 
2.46.0

