From 3709219aea8138c7ff6b6646b92fba30ecb5fa46 Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Sun, 10 Sep 2023 01:48:03 +0800
Subject: [PATCH] bus: mhi: pci_generic: add devices [1/2]

---
 drivers/bus/mhi/host/pci_generic.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/bus/mhi/host/pci_generic.c b/drivers/bus/mhi/host/pci_generic.c
index 336dba7..b874ad9 100644
--- a/drivers/bus/mhi/host/pci_generic.c
+++ b/drivers/bus/mhi/host/pci_generic.c
@@ -520,6 +520,8 @@ static const struct mhi_pci_dev_info mhi_telit_fn990_info = {
 };
 
 static const struct pci_device_id mhi_pci_id_table[] = {
+	{ PCI_DEVICE_SUB(PCI_VENDOR_ID_QCOM, 0x0306, PCI_VENDOR_ID_QCOM, 0x010c),
+		.driver_data = (kernel_ulong_t) &mhi_foxconn_sdx55_info },
 	/* EM919x (sdx55), use the same vid:pid as qcom-sdx55m */
 	{ PCI_DEVICE_SUB(PCI_VENDOR_ID_QCOM, 0x0306, 0x18d7, 0x0200),
 		.driver_data = (kernel_ulong_t) &mhi_sierra_em919x_info },
-- 
2.31.0

